apps/wing-console/console/ui/src/features/endpoints-pane/endpoint-tree.tsx

```diff
@@ -1,86 +1,125 @@
+  const expose = useCallback(
+    async (endpoint: EndpointItem) => {
+      await exposeEndpoint.mutateAsync({ resourcePath: endpoint.id });
+      showNotification(`${endpoint.label}`, {
+        type: "info",
+        body: "The endpoint is now exposed.",
       });
-      hideEndpoint(endpointId);
+    },
+    [exposeEndpoint, showNotification],
+  );
+
+  const onExposeEndpoint = useCallback(
+    async (endpoint: EndpointItem) => {
+      setSelectedEndpoint(endpoint);
+
+      if (showWarningModal) {
+        setWarningModalVisible(true);
+        return;
+      }
+      expose(endpoint);
+    },
+    [showWarningModal, expose],
+  );
+
+  const onHideEndpoint = useCallback(
+    async (endpoint: EndpointItem) => {
+      setSelectedEndpoint(endpoint);
+      await hideEndpoint.mutateAsync({ resourcePath: endpoint.id });
     },
     [hideEndpoint],
   );
 
+  const onAcceptWarning = useCallback(async () => {
+    if (!selectedEndpoint) {
+      return;
+    }
+    notifyWarningAccepted.mutate();
+    setShowWarningModal(false);
+    setWarningModalVisible(false);
+    expose(selectedEndpoint);
+  }, [notifyWarningAccepted, expose, selectedEndpoint]);
+
   useEffect(() => {
-    const isHidingEndpointDisconnected =
-      hidingEndpoint &&
-      endpointList.find(
-        (endpoint) =>
-          endpoint.exposeStatus === "disconnected" &&
-          hidingEndpoint.includes(endpoint.id),
-      );
-    if (isHidingEndpointDisconnected) {
-      setHidingEndpoint(undefined);
+    if (exposeEndpoint.error) {
+      showNotification(exposeEndpoint.error.message, { type: "error" });
     }
-  }, [endpointList, hidingEndpoint]);
-
-  const isLoading = useCallback(
-    (endpoint: EndpointItem) => {
-      return (
-        endpoint.exposeStatus === "connecting" ||
-        hidingEndpoint?.includes(endpoint.id)
-      );
-    },
-    [hidingEndpoint],
-  );
+  }, [exposeEndpoint.error, showNotification]);
+
+  const [initialNotification, setInitialNotification] = useState(true);
+
+  useEffect(() => {
+    if (!initialNotification) {
+      return;
+    }
+    if (endpointList.data) {
+      setInitialNotification(false);
+
+      const exposedEndpoints = endpointList.data
+        .filter((endpoint) => endpoint.exposeStatus === "connected")
+        .map((endpoint) => endpoint.label);
+
+      if (exposedEndpoints.length === 0) {
+        return;
+      }
+
+      showNotification("The following endpoints are exposed", {
+        body: exposedEndpoints.map((label) => <div key={label}>{label}</div>),
+        type: "info",
+        autoCloseDelayMs: 15_000,
+      });
+    }
+  }, [endpointList.data, showNotification, initialNotification]);
+
+  const { theme } = useTheme();
+
+  const endpointsSortedList = useMemo(() => {
+    return (
+      endpointList.data?.sort(
+        (a, b) =>
+          a.exposeStatus.localeCompare(b.exposeStatus) ||
+          a.label.localeCompare(b.label),
+      ) ?? []
+    );
+  }, [endpointList.data]);
```

Comment:
Careful here, the `Array.prototype.sort` mutates the array. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort.

Maybe we should clone before mutating.

```suggestion
  const endpointsSortedList = useMemo(() => {
    return (
      [...endpointList.data ?? []].sort(
        (a, b) =>
          a.exposeStatus.localeCompare(b.exposeStatus) ||
          a.label.localeCompare(b.label),
      )
    );
  }, [endpointList.data]);
```

Reply:
fixed 👍